template: fix KeyError: 'undefined variable: 0

zmedico opened the pull request:
Fixes: https://github.com/ansible/ansible/issues/20494

zmedico committed:
template: fix KeyError: 'undefined variable: 0

Fixes: https://github.com/ansible/ansible/issues/20494

abadger commented on lib/ansible/template/vars.py:
```diff
@@ -69,6 +71,14 @@ def __contains__(self, k):
+        return iter(keys)
+
+    def __len__(self):
+        return sum(1 for k in self.__iter__())
```
This will be a little slow (how apparent that is depends on how many variables a site has defined).  Even though it's not-DRY, Probably better to copy the ```__iter_```` code that creates the keys  here and then ```return len(keys)```

zmedico replied:
Okay, that's done in https://github.com/ansible/ansible/pull/27972. Thanks!

abadger commented:
@zmedico want to make this a PR against ansible/ansible?  jimi-c, bcoca, alikins, and I think this looks about right.

alikins committed:
intg tests for jinja-2.9 issues like 20494

test cases here are based on
https://github.com/ansible/ansible/issues/20494#issue-202108318

alikins committed:
add units/template/test_vars

alikins commented:
some intg tests the reproduce this at devel...alikins:jinja_29_intg_tests_20494

and some unit tests at https://github.com/ansible/ansible/compare/devel...alikins:jinja_29_20494_unit_tests

(separate branches so they can more easily be merged/picked into @zmedico branch)

zmedico commented:
Sent as https://github.com/ansible/ansible/pull/27972.
